src/Test/Utilities/Portable/Metadata/ILValidation.cs

```diff
@@ -3,16 +3,175 @@
+            {
+                var metadataReader = metadata.MetadataReader;
+                var peReader = metadata.Module.PEReaderOpt;
+                var peHeaders = peReader.PEHeaders;
+                var flags = peHeaders.CorHeader.Flags;
+
+                bool is32bit = peHeaders.PEHeader.Magic == PEMagic.PE32;
+                const int SectionHeaderSize = 40;
+                int peHeadersSize = peHeaders.PEHeaderStartOffset
+                    + PEHeaderSize(is32bit)
+                    + SectionHeaderSize * peHeaders.SectionHeaders.Length;
+
+                if (CorFlags.StrongNameSigned != (flags & CorFlags.StrongNameSigned))
+                {
+                    return false;
+                }
+
+                var snDirectory = peReader.PEHeaders.CorHeader.StrongNameSignatureDirectory;
+                int rva = snDirectory.RelativeVirtualAddress;
+                int size = snDirectory.Size;
+                var signature = peReader.GetSectionData(rva).GetContent(0, size);
+                if (!peHeaders.TryGetDirectoryOffset(snDirectory, out var snOffset))
+                {
+                    return false;
+                }
+
+                moduleContents.Position = 0;
+                int peSize = (int)moduleContents.Length;
+                var peImage = new BlobBuilder(peSize);
+                if (peSize != peImage.TryWriteBytes(moduleContents, peSize))
+                {
+                    return false;
+                }
+
+                byte[] buffer = GetBlobBuffer(peImage.GetBlobs().Single());
+
+                const int ChecksumOffset = 0x40;
+                uint expectedChecksum = peHeaders.PEHeader.CheckSum;
+                var checksumBlob = MakeBlob(buffer, peHeaders.PEHeaderStartOffset + ChecksumOffset, sizeof(uint));
+                var signatureBlob = MakeBlob(buffer, snOffset, size);
+
+                if (expectedChecksum != PeWriter.CalculateChecksum(peImage, checksumBlob))
+                {
+                    return false;
+                }
+
+                var snKey = CryptoBlobParser.ToRSAParameters(
+                    metadataReader.GetBlobBytes(metadataReader.GetAssemblyDefinition().PublicKey),
+                    includePrivateParameters: false);
+
+                // Signature is calculated with checksum zeroed
+                new BlobWriter(checksumBlob).WriteUInt32(0);
+
+                var content = GetContentToSign(peImage, peHeadersSize, peHeaders.PEHeader.FileAlignment, signatureBlob);
+                var hash = SigningUtilities.CalculateSha1(content);
+
+                var publicKey = CryptoBlobParser.ToRSAParameters(TestResources.General.snKey,
+                    includePrivateParameters: false);
+
+                using (var rsa = RSA.Create())
+                {
+                    rsa.ImportParameters(publicKey);
+                    var reversedSignature = signature.ToArray();
+                    Array.Reverse(reversedSignature);
```

Comment:
Can you add a comment on why we do a Reverse here? Given how much of a surprise it was to us may help to comment for future readers.

Reply:
I still have no idea why we do it -- I just know that the signature in core framework binaries is reversed.

Reply:
Understood. But even a comment to that effect will be helpful here. 

> The reason for reversing this signature is unknown. However because the original implementation had this reversal we must also as well. 

This gives a good accountability of "why" we made this change. Lacking the comment the next dev to read this will be wondering why in the same way we wonder why when we look at the CLR implementation 😄